src/mlir/design.md

```diff
@@ -0,0 +1,158 @@
+module every time for the functions and types that belong to them, to run module
+local optimisations.
+
+For this reason, we have chosen to create an MLIR `module` for each class/module,
+including the root module. For example:
+```
+// Verona code in directory FooBar
+class Foo { foo() { ... } }
+
+// Equivalent MLIR structure
+module @__ {
+  module @FooBar {
+    module @Foo {
+      func @foo() {
+        ...
+      }
+    }
+  }
+}
+```
+
+This structure, however, does not convert naturally to LLVM dialect. So we have
+a later pass that mangles the names and moves all functions to the root module
+just before lowering to LLVM dialect. The example above would be:
+
+```
+module @__ {
+  func @____FooBar__Foo__foo() {
+    ...
+  }
+}
+```
+
+And all calls to that function would also be changed in every other function that
+calls this one.
+
+## Load/Store Semantics
+
+Stack variables are allocated by LLVM IR with an `alloca` instruction. MLIR has
+a similar concept and we use that in the first implementation. The main problem
+with this approach is that there is only one dialect that has alloca/load/store:
+`memref`.
+
+MemRefs represent the memory used by `tensors`, which are multi-dimensional
+objects (matrices) of the same type. Verona is obviously not restricted to
+tensors as types, so we'll need to extend this to structure types, with complex
+layout semantics (ex. union types). For that, we'll need a set of dialect
+instructions that can be later lowered directly into LLVM for allocating stack
+space on complex layout semantics.
+
+For now, every function argument and local declaration (`let`, `var`) are
+allocated in the stack and its value stored in the memory space. This simplifies
+code generation (all variable symbols point to an address), but it doesn't help
+with value symbols and are really slow.
+
+The LLVM pipeline, even at O1, recognises most of the argument, temporary and
+local allocas as bogus and optimises them away, so this isn't a critical issue
+from a performance point of view, but we will probably need to update the symbol
+table soon to account for more types and states of the variables we hold (ex.
+`let` vs `var`).
+
+## Runtime Calls
+
+The Verona runtime is where the magic happens. Allocating regions, choosing the
+memory management strategy, scheduling behaviours, dispatching foreign language
+code, etc. The AST, however, has only the semantics of the language (with hints
+to the runtime behaviour), so we need to translate it at the MLIR lowering level.
+
+For example, heap allocation, which LLVM uses `malloc` calls, Verona uses
+`snmalloc`, which has a finer control over what we can do with the regions. In
+the end, however, it will be the same: calls to the runtime library in IR.
+
+The generator knows what runtime calls are appropriate for each AST construct,
+insert those calls in the right places and makes sure to declare the methods
+so that linking with the runtime library works.
```

Comment:
It would be good to discuss how the linking happens.  I presume the runtime will be compiled as IR and then made available as one of the inputs on a ThinLTO-like pipeline?

Reply:
I'd honestly rather leave that for later once we start implementing the first calls to the runtime. Also, this is not the compiler yet, and I imagine we'll have another module that will drive the parser, mlir and other libraries, including possibly a linking library (that will also look at foreign code, etc).

The fact that we can execute code at this early stage does not reflect on *how* we'll execute code in the final product.